# Content
1. Import Libraries
2. Import Data
3. Grouping and Aggregation TASK  

# 01 Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 02 Importing Data

In [2]:
path = r'C:\Users\tlayt\OneDrive\Documents\CF Data Analytics\Acheivement 4\Instacart Basket Analysis'

In [3]:
#import data
ords_prods_der = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_derived.pkl'))


# 03 Task

In [4]:
#grouping by depart_id and aggregating mean order number 
ords_prods_der.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

Q3 Department 21has the highest average orders where as department 5 has the lowest average orders. To the naked eye, these numbers rdo not varry very much howeer to further analyze their significance, a t-test could be performed. Overall the average orders range from 15.2 to 22.9. 
Additionaly, when looking at eh entroe data frame in comparison to the subset that only included a few departments, we get a more wholistic view and can see that department 17 and 19 are not the highest and lowest average orders. 

In [5]:
# steps 1 & 2. grouping by user id and aggregating max order number
ords_prods_der['max_order'] = ords_prods_der.groupby(['customer_id'])['order_number'].transform(np.max)

In [6]:
#checking
ords_prods_der.head(15)

order_id  customer_id  order_number  orders_day_of_week  \
0    2539329            1             1                   2   
1    2398795            1             2                   3   
2     473747            1             3                   3   
3    2254736            1             4                   4   
4     431534            1             5                   4   
5    3367565            1             6                   2   
6     550135            1             7                   1   
7    3108588            1             8                   1   
8    2295261            1             9                   1   
9    2550362            1            10                   4   
10   2968173           15            15                   1   
11   1870022           15            17                   2   
12   1911383           15            18                   2   
13   2715276           15            21                   1   
14    487368           15            22                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered product_name  aisle_id  department_id  prices _merge  \
0           0         Soda        77              7     9.0   both   
1           1         Soda        77              7     9.0   both   
2           1         Soda        77              7     9.0   both   
3           1         Soda        77              7     9.0   both   
4           1         Soda        77              7     9.0   both   
5           1         Soda        77              7     9.0   both   
6           1         Soda        77              7     9.0   both   
7           1         Soda        77              7     9.0   both   
8           1         Soda        77              7     9.0   both   
9           1         Soda        77              7     9.0   both   
10          0         Soda        77              7     9.0   both   
11          1         Soda        77              7     9.0   both   
12          1         Soda        77              7     9.0   both   
13          1         Soda        77              7     9.0   both   
14          1         Soda        77              7     9.0   both   

         price_labels     busiest_day        busiest_days  \
0   Mid-range product  Regularly busy      Regularly busy   
1   Mid-range product  Regularly busy  Least busiest days   
2   Mid-range product  Regularly busy  Least busiest days   
3   Mid-range product      Least busy  Least busiest days   
4   Mid-range product      Least busy  Least busiest days   
5   Mid-range product  Regularly busy      Regularly busy   
6   Mid-range product  Regularly busy       Busiests days   
7   Mid-range product  Regularly busy       Busiests days   
8   Mid-range product  Regularly busy       

In [8]:
#creating flag for loyalty customers 
ords_prods_der.loc[ords_prods_der['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [9]:
ords_prods_der.loc[(ords_prods_der['max_order'] <= 40) & (ords_prods_der['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [10]:
ords_prods_der.loc[ords_prods_der['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [12]:
ords_prods_der['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [16]:
#checking
ords_prods_der.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge       price_labels  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day        busiest_days busiest_period_of_day  max_order  \
0  Regularly busy      Regularly busy        Average Orders         10   
1  Regularly busy  Least busiest days        Average Orders         10   
2  Regularly busy  Least busiest days           Most Orders         10   
3      Least busy  Least busiest days        Average Orders         10   
4      Least busy  Least busiest days           Most Orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [18]:
# q5 multiple stats
ords_prods_der.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

In [19]:
# grouping by cust_id and finding mean price spent 
ords_prods_der.groupby('customer_id').agg({'prices':['mean']})

prices
                 mean
customer_id          
1            6.367797
2            7.515897
3            8.197727
4            8.205556
5            9.189189
...               ...
206205       8.909375
206206       7.646667
206207       7.313453
206208       8.366617
206209       7.058915

[206209 rows x 1 columns]

In [20]:
#creating column in df 
ords_prods_der['avg_spending'] = ords_prods_der.groupby(['customer_id'])['prices'].transform(np.mean)

In [21]:
#check
ords_prods_der.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge       price_labels  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      busiest_day        busiest_days busiest_period_of_day max_order  \
0  Regularly busy      Regularly busy        Average Orders        10   
1  Regularly busy  Least busiest days        Average Orders        10   
2  Regularly busy  Least busiest days           Most Orders        10   
3      Least busy  Least busiest days        Average Orders        10   
4      Least busy  Least busiest days           Most Orders        10   

   loyalty_flag avg_spending  
0  New customer     6.367797  
1  New customer     6.367797  
2  New customer     6.367797  
3  New customer     6.367797  
4  New customer     6.367797  

[5 rows x 21 columns]

In [22]:
#creating flag for spending 
ords_prods_der.loc[ords_prods_der['avg_spending'] < 10, 'spending_flag'] = 'Low spender'

In [23]:
ords_prods_der.loc[ords_prods_der['avg_spending'] > 10, 'spending_flag'] = 'High spender'

In [34]:
pd.options.display.max_rows = None

In [36]:
#check 
ords_prods_der.head(10)

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   
5   3367565            1             6                   2                  7   
6    550135            1             7                   1                  9   
7   3108588            1             8                   1                 14   
8   2295261            1             9                   1                 16   
9   2550362            1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   

  product_name  ...  prices  _merge       price_labels     busiest_day  \
0         Soda  ...     9.0    both  Mid-range product  Regularly busy   
1         Soda  ...     9.0    both  Mid-range product  Regularly busy   
2         Soda  ...     9.0    both  Mid-range product  Regularly busy   
3         Soda  ...     9.0    both  Mid-range product      Least busy   
4         Soda  ...     9.0    both  Mid-range product      Least busy   
5         Soda  ...     9.0    both  Mid-range product  Regularly busy   
6         Soda  ...     9.0    both  Mid-range product  Regularly busy   
7         Soda  ...     9.0    both  Mid-range product  Regularly busy   
8         Soda  ...     9.0    both  Mid-range product  Regularly busy   
9         Soda  ...     9.0    both  Mid-range product      Least busy   

         busiest_days busiest_period_of_day max_order  loyalty_flag  \
0      Regularly busy        Average Orders        10  New customer   
1  Least busiest days        Average Orders        10  New customer   
2  Least busiest days           Most Orders        10  New customer   
3  Least busiest days        Average Orders        10  New customer   
4  Least busiest days           Most Orders        10  New customer   
5      Regularly busy        Average Orders        10  New customer   
6       Busiests days        Average Orders        10  New customer   
7       Busiests days           Most Orders        10  New customer   
8       Busiests days           Most Orders        10  New customer   
9  Least busiest days        Average Orders        10  New customer   

   avg_spending spending_flag  
0      6.367797   Low spender  
1      6.367797   Low spender  
2      6.367797   Low spender  
3      6.367797   Low spender  
4      6.367797   Low spender  
5      6.367797   Low spender  
6      6.367797   Low spender  
7      6.367797   Low spender  
8      6.367797   Low spender  
9      6.367797   Low spender  

[10 rows x 22 columns]

In [30]:
#count
ords_prods_der['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      633864
NaN                  381
Name: spending_flag, dtype: int64

In [38]:
#resetting max rows 
pd.options.display.max_rows = 10

In [39]:
# grouping by cust_id and finding median of days since prior prder
ords_prods_der.groupby('customer_id').agg({'days_since_prior_order':['median']})

days_since_prior_order
                            median
customer_id                       
1                             20.5
2                             13.0
3                             10.0
4                             20.0
5                             11.0
...                            ...
206205                        30.0
206206                         3.0
206207                        16.0
206208                         7.0
206209                        22.0

[206209 rows x 1 columns]

In [41]:
#creating column in df 
ords_prods_der['med_DSPO'] = ords_prods_der.groupby(['customer_id'])['days_since_prior_order'].transform(np.median)

In [42]:
#check 
ords_prods_der.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge       price_labels     busiest_day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product  Regularly busy   
2         Soda  ...    both  Mid-range product  Regularly busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   

         busiest_days busiest_period_of_day max_order  loyalty_flag  \
0      Regularly busy        Average Orders        10  New customer   
1  Least busiest days        Average Orders        10  New customer   
2  Least busiest days           Most Orders        10  New customer   
3  Least busiest days        Average Orders        10  New customer   
4  Least busiest days           Most Orders        10  New customer   

  avg_spending  spending_flag med_DSPO  
0     6.367797    Low spender     20.5  
1     6.367797    Low spender     20.5  
2     6.367797    Low spender     20.5  
3     6.367797    Low spender     20.5  
4     6.367797    Low spender     20.5  

[5 rows x 23 columns]

In [43]:
#creating flag for freqency 
ords_prods_der.loc[ords_prods_der['med_DSPO'] > 20, 'freqency_flag'] = 'Non-frequent customer'

In [47]:
ords_prods_der.loc[(ords_prods_der['med_DSPO'] > 10) & (ords_prods_der['med_DSPO'] == 20), 'freqency_flag'] = 'Regular customer'

In [49]:
ords_prods_der.loc[ords_prods_der['med_DSPO'] <= 10, 'freqency_flag'] = 'Frequent customer'

In [51]:
ords_prods_der['freqency_flag'].value_counts(dropna = False)

Frequent customer        21559853
NaN                       6836355
Non-frequent customer     3636437
Regular customer           372214
Name: freqency_flag, dtype: int64

In [52]:
#oops. rename column
ords_prods_der.rename(columns = {'freqency_flag' : 'frequency_flag'}, inplace = True)

In [53]:
#checking 
ords_prods_der.head()

order_id  customer_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329            1             1                   2                  8   
1   2398795            1             2                   3                  7   
2    473747            1             3                   3                 12   
3   2254736            1             4                   4                  7   
4    431534            1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...       price_labels     busiest_day        busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy      Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy  Least busiest days   
2         Soda  ...  Mid-range product  Regularly busy  Least busiest days   
3         Soda  ...  Mid-range product      Least busy  Least busiest days   
4         Soda  ...  Mid-range product      Least busy  Least busiest days   

  busiest_period_of_day max_order  loyalty_flag avg_spending spending_flag  \
0        Average Orders        10  New customer     6.367797   Low spender   
1        Average Orders        10  New customer     6.367797   Low spender   
2           Most Orders        10  New customer     6.367797   Low spender   
3        Average Orders        10  New customer     6.367797   Low spender   
4           Most Orders        10  New customer     6.367797   Low spender   

   med_DSPO         frequency_flag  
0      20.5  Non-frequent customer  
1      20.5  Non-frequent customer  
2      20.5  Non-frequent customer  
3      20.5  Non-frequent customer  
4      20.5  Non-frequent customer  

[5 rows x 24 columns]

In [54]:
# Export data to pkl
ords_prods_der.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_agged.pkl'))
